### DISCUSSION TITLE
Action items from meetings

### DISCUSSION THREAD

**Discussion by asturur** [fabricjs/fabric.js#9485] [General]
** previous **

Layout manager

#9144 -> this i have to re read.

~~#9142 -> revisit and add activeSelection specs.~~

~~#9068 -> goes in with test, priority 1~~ done merged

~~#9043 -> i have to form an opinion.~~ moved to post release

~~#8950 -> request a line change - shachar put back the line.~~ done

~~Then after #8951 -> i will read it.~~ done

**6 november 2023**

exit editing on deselection:
https://github.com/fabricjs/fabric.js/pull/8950#discussion_r1199933126
Make proposal to handle this coherently across fabric, collect all the existing select/deselect side effects and plan how they could look like. -Shachar

~absolute: we merge as it is, we follow up with better math for the padding handling - Andrea~ done

Gradient is using isPath() we want to stop that and verify path, polygon, polyline work all same with a svg visual test: PR open https://github.com/fabricjs/fabric.js/pull/9488

24 November 2023
- merging #8951 surface some need for documenting add/remove in group and some semplification possible

**Comment by ShaMan123** [comment]
> exit editing on deselection:
> [#8950 (comment)](https://github.com/fabricjs/fabric.js/pull/8950#discussion_r1199933126)
> Make proposal to handle this coherently across fabric, collect all the existing select/deselect side effects and plan how they could look like. -Shachar

Insights from inspecting the code:
- `onSelect` has no side effects in the code
- `onDeselect` side effects can easily turn into a `deselected` event handler:
  - `IText` exits editing
  - `ActiveSelection` removes all objects


We should state in the jsdocs of these methods they should not contain any side effects

**Comment by asturur** [comment]
@ShaMan123 https://github.com/fabricjs/fabric.js/pull/9525/files figure out what is wrong or merge it.

**Reply by ShaMan123** [reply]
I commented there

**Comment by asturur** [comment]
@ShaMan123 take a decision/plan on the active selection reference on the canvas

**Reply by ShaMan123** [reply]
I looked into the code and I think it is best to change to ref concept.
I think the only change we need to do is to make the ref optional and clear it once the selection is discarded
That should be enough to satisfy our needs
We can remove the ref entirely and use `instance ActiveSelection` (since we import `ActiveSelection` in the canvas files there is not problem with imports)

<img width="1136" alt="Screenshot 2023-12-29 at 9 32 39" src="https://github.com/fabricjs/fabric.js/assets/34343793/88a24376-2ab7-4049-9788-1c3cfffef3e6">

**Reply by ShaMan123** [reply]
updated comment...

so actually we can remove the ref entirely and expose `Canvas#getActiveSelection` and that will replace all equality checks

```diff
- target === this._activeSelection
+ target === this.getActiveSelection()
```

```
getActiveSelection() {
   return this._activeObject instanceof ActiveSelection ? this._activeObject : undefined
}
```

**Reply by ShaMan123** [reply]
I did a draft #9561
We can discuss there

**Comment by asturur** [comment]
#9043 -> this is large and moved to post release. this is clean up internal state category

**Comment by asturur** [comment]
@ShaMan123  https://github.com/fabricjs/fabric.js/pull/9142 to write down what he wants to see improved in the interaction with active selections: if the changes come from a multiselect,a shift click, a drag select and so on.

**Reply by ShaMan123** [reply]
can be after the release
What is needed is context.
There are cases there is a need to know what triggered the object add/remove so custom logic can change the stack order for instance.
Maybe it is good enough to use the selection events

**Comment by asturur** [comment]
@asturur #9515 review and possibly merge

**Reply by asturur** [reply]
this is done

**Comment by asturur** [comment]
@asturur ~~reopen and merge this: https://github.com/fabricjs/fabric.js/pull/9470/~~ done

**Comment by ShaMan123** [comment]
we need to discuss async disposing as well before the release because of react

**Reply by asturur** [reply]
Disposing DOM is no more async and i don't think it interact with react badly.
Can you pull down here what do you think it needs to be discussed about it?

**Reply by ShaMan123** [reply]
I don't understand why you say it is not async.
#9554 states that react18 dev mode calls every useEffect twice to ensure it is cleaned up.
Since it is async at the second call fabric will throw the canvas is already initialized

**Reply by asturur** [reply]
As soon as canvas.dispose is called, our cleanup of the dom is sync.
If react run a piece of effect twice i hope it will call also the clean up function hopefully.
Whay is async is some later task we do on disposing the resoruces but has nothing to do with the dom cleanup.

```
  /**
   * Waits until rendering has settled to destroy the canvas
   * @returns {Promise<boolean>} a promise resolving to `true` once the canvas has been destroyed or to `false` if the canvas has was already destroyed
   * @throws if aborted by a consequent call
   */
  dispose() {
    !this.disposed &&
      this.elements.cleanupDOM({ width: this.width, height: this.height });
```

the part that matters to react, cleanupDOM is not async, this is what i mean.

**Reply by ShaMan123** [reply]
I understand now
So there should be no problem at all because I remember I added a test calling dispose more than once before it completes

**Reply by ShaMan123** [reply]
<img width="1512" alt="Screenshot 2024-01-07 at 8 18 33" src="https://github.com/fabricjs/fabric.js/assets/34343793/0858bd12-c320-4389-8bf9-4f42a9d8557f">

**Reply by ShaMan123** [reply]
Maybe what happens is that the dev is alarmed because `dispose` throws if it is called more than once.
Maybe we should change that?

**Comment by ShaMan123** [comment]
I think #8952 was forgotten

**Reply by asturur** [reply]
I think we can close this and replace isTextObject with something that doesn't use the type.
That refactor at this stage is off the table for me.
